SU2_CFD/src/solver_direct_mean.cpp

```diff
@@ -2273,23 +2280,55 @@ void CEulerSolver::Set_MPI_Solution_Gradient(CGeometry *geometry, CConfig *confi
-            Gradient[iVar][2] = rotMatrix[2][0]*Buffer_Receive_Gradient[0*nVar*nVertexR+iVar*nVertexR+iVertex] + rotMatrix[2][1]*Buffer_Receive_Gradient[1*nVar*nVertexR+iVar*nVertexR+iVertex] + rotMatrix[2][2]*Buffer_Receive_Gradient[2*nVar*nVertexR+iVar*nVertexR+iVertex];
-          }
-        }
+        for (iVar = 0; iVar < nVar; iVar++){
+          for (iDim = 0; iDim < nDim; iDim++){
+            //Gradient[iVar][iDim] = Buffer_Receive_Gradient[iDim*nVar*nVertexR+iVar*nVertexR+iVertex];
+	    int index = iDim*nVar*nVertexR+iVar*nVertexR+iVertex;
+	    GradientTmp[iVar][iDim] = Buffer_Receive_Gradient[index];
+	  }
+	}
         
+        // /*--- Need to rotate the gradients for all conserved variables. ---*/
+        // for (iVar = 0; iVar < nVar; iVar++) {
+        //   if (nDim == 2) {
+        //     Gradient[iVar][0] = rotMatrix[0][0]*Buffer_Receive_Gradient[0*nVar*nVertexR+iVar*nVertexR+iVertex] + rotMatrix[0][1]*Buffer_Receive_Gradient[1*nVar*nVertexR+iVar*nVertexR+iVertex];
+        //     Gradient[iVar][1] = rotMatrix[1][0]*Buffer_Receive_Gradient[0*nVar*nVertexR+iVar*nVertexR+iVertex] + rotMatrix[1][1]*Buffer_Receive_Gradient[1*nVar*nVertexR+iVar*nVertexR+iVertex];
+        //   }
+        //   else {
+        //     Gradient[iVar][0] = rotMatrix[0][0]*Buffer_Receive_Gradient[0*nVar*nVertexR+iVar*nVertexR+iVertex] + rotMatrix[0][1]*Buffer_Receive_Gradient[1*nVar*nVertexR+iVar*nVertexR+iVertex] + rotMatrix[0][2]*Buffer_Receive_Gradient[2*nVar*nVertexR+iVar*nVertexR+iVertex];
+        //     Gradient[iVar][1] = rotMatrix[1][0]*Buffer_Receive_Gradient[0*nVar*nVertexR+iVar*nVertexR+iVertex] + rotMatrix[1][1]*Buffer_Receive_Gradient[1*nVar*nVertexR+iVar*nVertexR+iVertex] + rotMatrix[1][2]*Buffer_Receive_Gradient[2*nVar*nVertexR+iVar*nVertexR+iVertex];
+        //     Gradient[iVar][2] = rotMatrix[2][0]*Buffer_Receive_Gradient[0*nVar*nVertexR+iVar*nVertexR+iVertex] + rotMatrix[2][1]*Buffer_Receive_Gradient[1*nVar*nVertexR+iVar*nVertexR+iVertex] + rotMatrix[2][2]*Buffer_Receive_Gradient[2*nVar*nVertexR+iVar*nVertexR+iVertex];
+        //   }
+        // }
+
+        /*--- Rotate the gradients for all SCALAR conserved variables. ---*/
+	for (iDim=0; iDim<nDim; iDim++) {
+	  Gradient[0][iDim] = Gradient[nDim+1][iDim] = 0;
+	  for (jDim=0; jDim<nDim; jDim++) {
+	    // Density iVar = 0
+	    Gradient[0][iDim] += rotMatrix[iDim][jDim]*GradientTmp[0][jDim];
+
+	    // Energy iVar = nDim+1
+	    Gradient[nDim+1][iDim] += rotMatrix[iDim][jDim]*GradientTmp[nDim+1][jDim];
+	  }
+	}
+	 
+        /*--- Rotate the gradients for momentum gradient (a tensor). ---*/
+	for (iDim=0; iDim<nDim; iDim++) {
+	  for (jDim=0; jDim<nDim; jDim++) {
+	    // iDim momentum component gradient wrt jDim direction
+	    Gradient[iDim+1][jDim] = 0;
+
+	    for (kDim=0; kDim<nDim; kDim++) {
+	      for (lDim=0; lDim<nDim; lDim++) {
+		Gradient[iDim+1][jDim] +=
+		  rotMatrix[iDim][kDim]*GradientTmp[kDim+1][lDim]*rotMatrix[jDim][lDim];
+	      }
+	    }
+	  }
+	}
+
+
```

Comment:
Is this a hack, or is it an improvement? Should it stay in the code?

Reply:
This is intended to be an improvement.  The old code transforms the velocity gradient as if it were a vector not a tensor, which is wrong.  The new version is intended to fix that.  But, it hasn't been tested much, so it could still be buggy.